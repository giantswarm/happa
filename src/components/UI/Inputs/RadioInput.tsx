import styled from '@emotion/styled';
import PropTypes from 'prop-types';
import React, { ComponentProps } from 'react';

const Label = styled.label`
  cursor: pointer;
  display: flex;
  align-items: center;
`;

const StyledInput = styled.input`
  visibility: hidden;
  opacity: 0;
`;

const Bullet = styled.span<{}>`
  display: block;
  position: relative;
  border: ${({ theme }) => theme.border};
  background: ${({ theme }) => theme.colors.white1};
  border-radius: 50%;
  vertical-align: middle;
  width: 16px;
  height: 16px;
  z-index: 0;
  margin-right: 8px;

  &:after {
    content: '';
    position: absolute;
    left: calc(4px / 2);
    top: 0;
    bottom: 0;
    margin: auto;
    background: ${({ theme }) => theme.colors.shade2};
    border-radius: 50%;
    width: calc(100% - 4px);
    height: calc(100% - 4px);
    z-index: 1;

    transition: 0.4s cubic-bezier(0.81, 0, 0.34, 1.75);
    transform: scale(0.01);
    opacity: 0;
  }

  ${StyledInput}:checked + &:after {
    opacity: 1;
    transform: scale(1);
  }
`;

const LabelText = styled.span`
  font-weight: 300;
`;

interface IRadioInputProps extends ComponentProps<'input'> {
  label?: string;
}

const RadioInput: React.FC<IRadioInputProps> = ({ id, label, ...rest }) => {
  return (
    <Label htmlFor={id}>
      <StyledInput {...rest} type='radio' id={id} />
      <Bullet />
      <LabelText>{label}</LabelText>
    </Label>
  );
};

RadioInput.propTypes = {
  id: PropTypes.string,
  label: PropTypes.string,
};

export default RadioInput;
