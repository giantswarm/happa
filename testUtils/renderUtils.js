import { render } from '@testing-library/react';
import App from 'App';
import { ConnectedRouter } from 'connected-react-router';
import { ThemeProvider } from 'emotion-theming';
import { createMemoryHistory } from 'history';
import React from 'react';
import { Provider } from 'react-redux';
import { AppRoutes } from 'shared/constants/routes';
import configureStore from 'stores/configureStore';
import theme from 'styles/theme';

export const initialStorage = {
  user:
    '{"email":"developer@giantswarm.io","auth":{"scheme":"giantswarm","token":"a-valid-token"},"isAdmin":true}',
};

/**
 * This function will render the whole app with a mocked store in the route
 * provided.
 * @param {String} route.
 * @param {HTMLElement} container An HTMLElement.
 */
export function renderRouteWithStore(
  initialRoute = AppRoutes.Home,
  state = {},
  storage = initialStorage
) {
  localStorage.replaceWith(storage);

  const history = createMemoryHistory({
    initialEntries: [initialRoute],
    initialIndex: 0,
  });

  const store = configureStore(state, history);

  const app = render(<App {...{ store, theme, history }} />);

  return app;
}

/**
 * Render component in the Theme context
 * @param {React.ReactType} component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {<Q extends TestingLibrary.Queries>TestingLibrary.RenderOptions<Q>} [options] Testing library render options
 */
export function renderWithTheme(component, props, options) {
  return render(getComponentWithTheme(component, props), options);
}

/**
 * Include component in the Theme context and return it
 * @param {React.ReactType} Component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 */
export function getComponentWithTheme(Component, props) {
  return (
    <ThemeProvider theme={theme}>
      <Component {...props} />
    </ThemeProvider>
  );
}

/**
 * Render component in the Store context
 * @param {React.ReactType} component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {Record<string, any>} state Current Store state
 * @param {<Q extends TestingLibrary.Queries>TestingLibrary.RenderOptions<Q>} [options] Testing library render options
 */
export function renderWithStore(component, props, state, options) {
  return render(getComponentWithStore(component, props, state), options);
}

/**
 * Include component in the Store context and return it
 * @param {React.ReactType} Component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {Record<string, any>} state Current Store state
 * @param {Record<string, any>} storage Current LocalStorage value
 * @param {History} history Current Browser history
 */
export function getComponentWithStore(
  Component,
  props = {},
  state = {},
  storage = initialStorage,
  history = createMemoryHistory()
) {
  localStorage.replaceWith(storage);

  const store = configureStore(state, history);

  const app = (
    <Provider store={store}>
      <ThemeProvider theme={theme}>
        <ConnectedRouter history={history}>
          <Component {...props} />
        </ConnectedRouter>
      </ThemeProvider>
    </Provider>
  );

  return app;
}
