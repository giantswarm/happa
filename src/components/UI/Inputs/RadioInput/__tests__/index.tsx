import { renderWithTheme } from 'test/renderUtils';

import RadioInput from '..';

describe('RadioInput', () => {
  it('renders a simple input', () => {
    const { container } = renderWithTheme(RadioInput, {
      name: 'some-input',
      label: 'Some input',
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders an input with a form field label', () => {
    const { container } = renderWithTheme(RadioInput, {
      name: 'some-input',
      label: 'Some input',
      fieldLabel: 'A brilliant input',
      id: 'some-input',
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders an input with a validation error', () => {
    const { container } = renderWithTheme(RadioInput, {
      name: 'some-input',
      label: 'Some input',
      error: 'You failed at this',
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders an input with a help message', () => {
    const { container } = renderWithTheme(RadioInput, {
      name: 'some-input',
      label: 'Some input',
      help: 'This is a very cool input',
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders an input with an info message', () => {
    const { container } = renderWithTheme(RadioInput, {
      name: 'some-input',
      label: 'Some input',
      info: 'You can set a cool value',
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders a disabled input', () => {
    const { container } = renderWithTheme(RadioInput, {
      name: 'some-input',
      label: 'Some input',
      disabled: true,
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});
