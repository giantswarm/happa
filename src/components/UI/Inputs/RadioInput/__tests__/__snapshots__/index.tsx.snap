// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioInput renders a disabled input 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 fxRgDK"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 lixbyN"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 jOmtjA FormField__FormFieldBox-sc-m9hood-0 jTOPYT RadioInput__StyledFormField-sc-1brmdkl-0 btSJIi"
      tabindex="0"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 a-DZpO FormField__FormFieldContentBox-sc-m9hood-1"
      >
        <label
          class="StyledRadioButton__StyledRadioButtonContainer-sc-g1f6ld-0 bymbLY"
          disabled=""
        >
          <div
            class="StyledBox-sc-13pk1d4-0 cqzCLd StyledRadioButton-sc-g1f6ld-5 hXfoNY"
          >
            <input
              class="StyledRadioButton__StyledRadioButtonInput-sc-g1f6ld-1 dTlTkb"
              disabled=""
              name="some-input"
              tabindex="-1"
              type="radio"
            />
            <div
              class="StyledBox-sc-13pk1d4-0 eOIUqD StyledRadioButton__StyledRadioButtonBox-sc-g1f6ld-4 hZSOaX"
            />
          </div>
          <span
            class="StyledRadioButton__StyledRadioButtonLabel-sc-g1f6ld-2 fnudST"
          >
            Some input
          </span>
        </label>
      </div>
    </div>
  </div>
</div>
`;

exports[`RadioInput renders a simple input 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 fxRgDK"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 lixbyN"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 jOmtjA FormField__FormFieldBox-sc-m9hood-0 jTOPYT RadioInput__StyledFormField-sc-1brmdkl-0 btSJIi"
      tabindex="0"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 a-DZpO FormField__FormFieldContentBox-sc-m9hood-1"
      >
        <label
          class="StyledRadioButton__StyledRadioButtonContainer-sc-g1f6ld-0 hZkGAh"
        >
          <div
            class="StyledBox-sc-13pk1d4-0 cqzCLd StyledRadioButton-sc-g1f6ld-5 hXfoNY"
          >
            <input
              class="StyledRadioButton__StyledRadioButtonInput-sc-g1f6ld-1 cpHFwb"
              name="some-input"
              tabindex="-1"
              type="radio"
            />
            <div
              class="StyledBox-sc-13pk1d4-0 eOIUqD StyledRadioButton__StyledRadioButtonBox-sc-g1f6ld-4 hZSOaX"
            />
          </div>
          <span
            class="StyledRadioButton__StyledRadioButtonLabel-sc-g1f6ld-2 fnudST"
          >
            Some input
          </span>
        </label>
      </div>
    </div>
  </div>
</div>
`;

exports[`RadioInput renders an input with a form field label 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 fxRgDK"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 lixbyN"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 jOmtjA FormField__FormFieldBox-sc-m9hood-0 jTOPYT RadioInput__StyledFormField-sc-1brmdkl-0 btSJIi"
      tabindex="0"
    >
      <label
        class="StyledText-sc-1sadyjn-0 imNpig"
        for="some-input"
      >
        A brilliant input
      </label>
      <div
        class="StyledBox-sc-13pk1d4-0 a-DZpO FormField__FormFieldContentBox-sc-m9hood-1"
      >
        <label
          class="StyledRadioButton__StyledRadioButtonContainer-sc-g1f6ld-0 hZkGAh"
          for="some-input"
        >
          <div
            class="StyledBox-sc-13pk1d4-0 cqzCLd StyledRadioButton-sc-g1f6ld-5 hXfoNY"
          >
            <input
              class="StyledRadioButton__StyledRadioButtonInput-sc-g1f6ld-1 cpHFwb"
              id="some-input"
              name="some-input"
              tabindex="-1"
              type="radio"
            />
            <div
              class="StyledBox-sc-13pk1d4-0 eOIUqD StyledRadioButton__StyledRadioButtonBox-sc-g1f6ld-4 hZSOaX"
            />
          </div>
          <span
            class="StyledRadioButton__StyledRadioButtonLabel-sc-g1f6ld-2 fnudST"
          >
            Some input
          </span>
        </label>
      </div>
    </div>
  </div>
</div>
`;

exports[`RadioInput renders an input with a help message 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 fxRgDK"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 lixbyN"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 jOmtjA FormField__FormFieldBox-sc-m9hood-0 jTOPYT RadioInput__StyledFormField-sc-1brmdkl-0 btSJIi"
      tabindex="0"
    >
      <span
        class="StyledText-sc-1sadyjn-0 dXxgdw"
      >
        This is a very cool input
      </span>
      <div
        class="StyledBox-sc-13pk1d4-0 a-DZpO FormField__FormFieldContentBox-sc-m9hood-1"
      >
        <label
          class="StyledRadioButton__StyledRadioButtonContainer-sc-g1f6ld-0 hZkGAh"
        >
          <div
            class="StyledBox-sc-13pk1d4-0 cqzCLd StyledRadioButton-sc-g1f6ld-5 hXfoNY"
          >
            <input
              class="StyledRadioButton__StyledRadioButtonInput-sc-g1f6ld-1 cpHFwb"
              name="some-input"
              tabindex="-1"
              type="radio"
            />
            <div
              class="StyledBox-sc-13pk1d4-0 eOIUqD StyledRadioButton__StyledRadioButtonBox-sc-g1f6ld-4 hZSOaX"
            />
          </div>
          <span
            class="StyledRadioButton__StyledRadioButtonLabel-sc-g1f6ld-2 fnudST"
          >
            Some input
          </span>
        </label>
      </div>
    </div>
  </div>
</div>
`;

exports[`RadioInput renders an input with a validation error 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 fxRgDK"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 lixbyN"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 jOmtjA FormField__FormFieldBox-sc-m9hood-0 jTOPYT RadioInput__StyledFormField-sc-1brmdkl-0 btSJIi"
      tabindex="0"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 a-DZpO FormField__FormFieldContentBox-sc-m9hood-1"
      >
        <label
          class="StyledRadioButton__StyledRadioButtonContainer-sc-g1f6ld-0 hZkGAh"
        >
          <div
            class="StyledBox-sc-13pk1d4-0 cqzCLd StyledRadioButton-sc-g1f6ld-5 hXfoNY"
          >
            <input
              class="StyledRadioButton__StyledRadioButtonInput-sc-g1f6ld-1 cpHFwb"
              name="some-input"
              tabindex="-1"
              type="radio"
            />
            <div
              class="StyledBox-sc-13pk1d4-0 eOIUqD StyledRadioButton__StyledRadioButtonBox-sc-g1f6ld-4 hZSOaX"
            />
          </div>
          <span
            class="StyledRadioButton__StyledRadioButtonLabel-sc-g1f6ld-2 fnudST"
          >
            Some input
          </span>
        </label>
      </div>
      <span
        class="StyledText-sc-1sadyjn-0 dGdtNa"
      >
        You failed at this
      </span>
    </div>
  </div>
</div>
`;

exports[`RadioInput renders an input with an info message 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 fxRgDK"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 lixbyN"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 jOmtjA FormField__FormFieldBox-sc-m9hood-0 jTOPYT RadioInput__StyledFormField-sc-1brmdkl-0 btSJIi"
      tabindex="0"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 a-DZpO FormField__FormFieldContentBox-sc-m9hood-1"
      >
        <label
          class="StyledRadioButton__StyledRadioButtonContainer-sc-g1f6ld-0 hZkGAh"
        >
          <div
            class="StyledBox-sc-13pk1d4-0 cqzCLd StyledRadioButton-sc-g1f6ld-5 hXfoNY"
          >
            <input
              class="StyledRadioButton__StyledRadioButtonInput-sc-g1f6ld-1 cpHFwb"
              name="some-input"
              tabindex="-1"
              type="radio"
            />
            <div
              class="StyledBox-sc-13pk1d4-0 eOIUqD StyledRadioButton__StyledRadioButtonBox-sc-g1f6ld-4 hZSOaX"
            />
          </div>
          <span
            class="StyledRadioButton__StyledRadioButtonLabel-sc-g1f6ld-2 fnudST"
          >
            Some input
          </span>
        </label>
      </div>
      <span
        class="StyledText-sc-1sadyjn-0 gLxcNH"
      >
        You can set a cool value
      </span>
    </div>
  </div>
</div>
`;
